libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -133,52 +163,217 @@ void ContractLevelChecker::findDuplicateDefinitions(map<string, vector<T>> const
 	}
 }
 
+vector<ContractDefinition const*> resolveBaseContracts(ContractDefinition const& _contract)
+{
+	vector<ContractDefinition const*> resolvedContracts{_contract.baseContracts().size(), nullptr};
+
+	transform(
+		_contract.baseContracts().cbegin(),
+		_contract.baseContracts().cend(),
+		resolvedContracts.begin(),
+		[] (ASTPointer<InheritanceSpecifier> const& _specifier)
+		{
+			Declaration const* baseDecl =
+				_specifier->name().annotation().referencedDeclaration;
+			return dynamic_cast<ContractDefinition const*>(baseDecl);
+		}
+	);
+
+	return resolvedContracts;
+}
+
+ContractLevelChecker::FunctionInfo ContractLevelChecker::findAndCheckOverride(
+	ContractDefinition const& _derived,
+	FunctionDefinition const& _function,
+	ContractDefinition const& _base
+)
+{
+	// Look for possible override in all functions of _base
+	for (FunctionDefinition const* baseFunc: _base.definedFunctions())
+	{
+		if (!hasEqualNameAndParameters(_function, *baseFunc))
+			continue;
+
+		if (!checkFunctionOverride(_function, *baseFunc))
+			return {};
+
+		return {baseFunc, &_base};
+	}
+
+	// If we found no matching override, check the bases of base
+	for (auto const c: resolveBaseContracts(_base))
+	{
+		auto const& baseBase = findAndCheckOverride(_derived, _function, *c);
+
+		if (baseBase.funcDef)
+			return baseBase;
+	}
+
+	return {};
+}
+
 void ContractLevelChecker::checkIllegalOverrides(ContractDefinition const& _contract)
 {
+	for (FunctionDefinition const* function: _contract.definedFunctions())
```

Comment:
Maybe move this into its own function (or at least the for loop body) or call `checkAmbiguousOverrides` directly from `ContractLevelChecker::check`

Reply:
The function is still too long, please split it up.